keps/sig-network/2433-topology-aware-hints/README.md

```diff
@@ -0,0 +1,578 @@
+of the following scenarios:
+
+1. A deleted Node results in a Service exceeding the overload threshold.
+2. A new Node results in a Service that is able to achieve an endpoint
+   distribution below 20% for the first time.
+
+### Future Expansion
+
+In the future we may expand this functionality if needed. This could include:
+
+- A new `RequireZone` algorithm that would keep endpoints in EndpointSlices for
+  the same zone they are in.
+- A new option to specify a minimum threshold for the `Auto` approach.
+- Support for region based hints.
+
+### Test Plan
+
+#### Controller Unit Tests
+| Test Description | Expected Result |
+| :--- | :--- |
+| Feature Gate On, Annotation == 'Auto', 2+ zones | Hints set |
+| Feature Gate On, Annotation == 'Auto', 1 zone | No hints set |
+| Feature Gate On, Annotation == 'On', 2+ zones | No hints |
+| Feature Gate On, Annotation Off, 2+ zones | No hints |
+| Feature Gate Off, Annotation On, 2+ zones | No hints |
+| Feature Gate Off, Annotation Off, 2+ zones | No hints |
+| Feature Gate Off, Annotation Off, 2+ zones | No hints |
+| 2 endpoints, 3 zones | No hints |
+| 3 endpoints, 3 zones | Hints set |
+| 4 endpoints, 3 zones | No hints |
+| 4 endpoints, 2 zones | Hints set |
+| 4 endpoints all from 1 zone, 2 zones | Hints set |
+| 4 endpoints, 3 zones, 1 zone with 2x cores | Hints set |
+| 400 endpoints, 4 zones with slightly different cores | Hints set |
+| Node removal that does not trigger threshold transition | No EndpointSlice changes |
+| Node removal that triggers threshold transition | EndpointSlice updates |
+| Node without way to determine cores | All Nodes treated equally |
+| Endpoint additions that require redistribution | Hints updated |
+| Endpoint removals that require redistribution | Hints updated |
+
+#### Kube-Proxy Unit Tests
+| Test Description | Expected Result |
+| :--- | :--- |
+| Feature Gate On, Annotation == 'Auto', hints matching zone | Endpoints filtered |
+| Feature Gate On, Annotation == 'On', hints matching zone | Endpoints not filtered |
+| Feature Gate Off, Annotation == 'Auto', hints matching zone | Endpoints not filtered |
+| Feature Gate On, Annotation == 'Auto', no hints matching zone | Endpoints not filtered |
+
+### Observability
+We can reuse some of the metrics of EndpointSlice Controller that we already
+have in the current version to observe the changes of endpoints (addition,
+deletion and update). Meanwhile we can add more metrics to have a glimpse of
+different approaches.
+
+- `endpoint_slice_controller/endpointslices_changed_per_sync`
+- `endpoint_slice_controller/syncs`
+
+```
+const SubSystem = "endpoint_slice_controller"
+
+// This metric observes churn of EndpointSlices per sync
+EPSChangedPerSync = metrics.NewHistogramVec(
+  &metrics.HistogramOpts{
+	  Subsystem: Subsystem,
+    Name: "endpointslices_changed_per_sync",
+    Help: "Number of EndpointSlices be changed on each Service sync",
+  },
+  []string{"approach"}, // either "random" or "auto"
```

Comment:
maybe adding "zone" too?

Reply:
I think this would only make sense with the previous proposal based on subsetting. In this case, endpoints from and for different zones will all be represented in the same slices.

Reply:
Added new metric that should cover this.